mws/apis/easyship.py

```diff
@@ -0,0 +1,83 @@
+                          package_height=0, package_length=0, package_dimensions_uom='cm',
+                          package_weight=0, package_weight_uom='g'):
+        data = {
+            'Action': 'ListPickupSlots',
+            'MarketplaceId': marketplace_id,
+            'AmazonOrderId': amazon_order_id,
+            'PackageDimensions.Width': str(package_width),
+            'PackageDimensions.Length': str(package_length),
+            'PackageDimensions.Height': str(package_height),
+            'PackageDimensions.Unit': package_dimensions_uom,
+            'PackageWeight.Unit': package_weight_uom,
+            'PackageWeight.Value': str(package_weight)
+        }
+        return self.make_request(data)
+
+    def create_scheduled_package(self, marketplace_id=None, amazon_order_id=None, package_width=0,
+                                 package_height=0, package_length=0, package_dimensions_uom='cm',
+                                 package_weight=0, package_weight_uom='g',
+                                 slot_id=None, slot_start_time=None,
+                                 slot_end_time=None, package_identifier=None):
+        data = {
+            'Action': 'CreateScheduledPackage',
+            'MarketplaceId': marketplace_id,
+            'AmazonOrderId': amazon_order_id,
+            'PackageRequestDetails.PackageDimensions.Length': str(package_length),
+            'PackageRequestDetails.PackageDimensions.Width': str(package_width),
+            'PackageRequestDetails.PackageDimensions.Height': str(package_height),
+            'PackageRequestDetails.PackageDimensions.Unit': package_dimensions_uom,
+            'PackageRequestDetails.PackageWeight.Unit': package_weight_uom,
+            'PackageRequestDetails.PackageWeight.Value': str(package_weight),
+            'PackageRequestDetails.PackagePickupSlot.SlotId': slot_id,
+            'PackageRequestDetails.PackagePickupSlot.PickupTimeStart': slot_start_time,
+            'PackageRequestDetails.PackagePickupSlot.PickupTimeEnd': slot_end_time,
+            'PackageRequestDetails.PackageIdentifier': package_identifier
+
+        }
+        return self.make_request(data)
+
+    def update_scheduled_packages(self, marketplace_id=None, package_update_details=None):
+        data = {
+            'Action': 'UpdateScheduledPackages',
+            'MarketplaceId': marketplace_id
+        }
+        package_update_details = package_update_details or []
+        for detail in package_update_details:
+            detail['ScheduledPackageId.AmazonOrderId'] = detail.pop('amazon_order_id')
+            detail['ScheduledPackageId.PackageId'] = detail.pop('package_id')
+            detail['PackagePickupSlot.SlotId'] = detail.pop('slot_id')
+            detail['PackagePickupSlot.PickupTimeStart'] = detail.pop('slot_start_time', None)
+            detail['PackagePickupSlot.PickupTimeEnd'] = detail.pop('slot_end_time', None)
```

Comment:
I have minor doubts about this particular pattern, but not a showstopper. Changes will be accepted, and I encourage you to contribute an additional change with the following.

---

To be clear, this code does work as intended, even in a case where it doesn't seem like it should. For instance, if `package_update_details` is an immutable sequence, like a tuple, one might think it can't be updated. In fact, the dicts contained in the sequence can be updated, and the tuple that points to them is not being changed. Though this works, I would like to see it cleaned up so that we aren't changing an object as we iterate through it.

Further, if the `detail` inside the loop is not a dict, it will throw `TypeError: string indices must be integers` when trying to access one of its keys. That's not a clear error message for what actually happened, so I would like to see that error handled in a way that contextualizes the problem for the end user. For example:

```python
if not isinstance(detail, dict):
    raise ValueError("Package details must be a list of dicts, each with required keys `amazon_order_id`, `package_id`, and `slot_id` (optional keys: `slot_start_time` and `slot_end_time`)")
```

Similarly, `detail.pop('amazon_order_id')` will throw `KeyError: 'amazon_order_id'` if that key is missing. Again, in context, this is a bit obscure, and we want the user to see some more details to help them out. Ideally, the dict should be tested for those required keys, with the same `ValueError` as above thrown if any are missing, so that the developer knows what went wrong more quickly.

These concerns can be handled by a utility function from elsewhere in the module.

Reply:
@GriceTurrble thanks for pointing this out. I will fix these points. The `package_update_details` won't be modified in the function. And the function will throw an error if a non dict or a dict without required keys is passed.